src/mlpack/core/tree/binary_space_tree/binary_space_tree.hpp

```diff
@@ -333,6 +333,36 @@ class BinarySpaceTree
   size_t NumChildren() const;
 
   /**
+   * Return the nearest child node to the given query point.  If this is a leaf
+   * node, it will return a reference to itself.
+   */
+  template<typename VecType>
+  BinarySpaceTree& GetNearestChild(
+      const VecType& point,
+      typename boost::enable_if<IsVector<VecType> >::type* = 0);
+
+  /**
+   * Return the furthest child node to the given query point.  If this is a leaf
+   * node, it will return a reference to itself.
+   */
+  template<typename VecType>
+  BinarySpaceTree& GetFurthestChild(
+      const VecType& point,
+      typename boost::enable_if<IsVector<VecType> >::type* = 0);
+
+  /**
+   * Return the nearest child node to the given query node.  If it can't decide
+   * it will return a null pointer.
+   */
+  BinarySpaceTree* GetNearestChild(const BinarySpaceTree& queryNode);
+
+  /**
+   * Return the furthest child node to the given query node.  If it can't decide
+   * it will return a null pointer.
+   */
+  BinarySpaceTree* GetFurthestChild(const BinarySpaceTree& queryNode);
```

Comment:
@rcurtin  The overloads of spill trees are used for `DefeatistDualTreeTraverser`. Also, all of these overloads will be used when implementing a greedy dual tree traverser.
I agree that it is best to return an index. I have updated the code in: https://github.com/mlpack/mlpack/pull/762/commits/8e305125edbc85f6f037fbb01b081509153641ff
Thanks!